### DISCUSSION TITLE
I still don't get the differences between `@babel/preset-env`.`useBuiltIns: entry` and `useBuiltIns: usage`

### DISCUSSION THREAD

**Discussion by nyngwang** [babel/babel#14443] [Q&A] [answered]
Under the section of `useBuiltIns: entry`:

> Importing "core-js" loads polyfills for every possible ECMAScript feature: what if you know that you only need some of them? When using core-js@3, @babel/preset-env is able to **optimize** every single core-js entrypoint and their combinations. For example, you might want to only polyfill array methods and new Math proposals: [...]

So this sounds like it is doing the same thing as `useBuiltIns: usage`. But with the latter option, the result bundle size is (much) smaller. To be more specific, I added the (only) following line at the entrypoint:

```js
import 'core-js'
```

and for testing, I only used 1 single feature that needs to be polyfills. But with `useBuiltIns: entry`, I got this debug message from my terminal:

```
The corejs3 polyfill entry has been replaced with the following polyfills:
  es.symbol { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  es.symbol.description { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  es.symbol.async-iterator { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }

  // many lines omitted

  web.queue-microtask { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  web.structured-clone { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  web.url { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  web.url.to-json { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
  web.url-search-params { "android":"99", "chrome":"97", "edge":"98", "firefox":"97", "ie":"11", "ios":"12.2", "opera":"82", "safari":"14.1", "samsung":"15" }
```

But I'm pretty sure that I didn't use things like `es.typed_array` in my code. Is this something you called *optimize* in the quote?

> [...] When using core-js@3, @babel/preset-env is able to **optimize** every single core-js entrypoint and their combinations. [...]

Any explain?

**Comment by nicolo-ribaudo** [comment, marked as answer]
There are two different "filters":
1. Which features are supported by your target browsers
2. Which features you are using

`entry` only considers (1) (thus it optimizes `import "core-js"` to the minimum amount of imports needed for your targets), while `usage` also considers (2).

**Reply by nyngwang** [reply]
@nicolo-ribaudo: OK, now the problem: No one will use `usage` because [someone in the old thread](https://github.com/babel/babel/issues/9853#issuecomment-619587386) was saying that with `usage` people "would" miss some polyfills used in library or helpers? So what's the best practice right now in 2022?

**Reply by nicolo-ribaudo** [reply]
Yes, it's not possible to statically detect which polyfills are needed.

Consider this example:
```js
const name = prompt("What array method do you want to know the length of?");
alert("It's " + [][name].length);
```
if the users writes `flatMap` in the prompt opened by the browser, you will need the `flatMap` polyfill. However, it's not possible for Babel to know it statically, so you can tell it to load everything with `useBuiltIns: "entry"`. Additionally, Babel cannot know which polyfills are needed in files that you are not processing with Babel (if you are only processing part of your project).

My recommendation is to use `useBuiltIns: "usage"` if you are ensuring that your code works in old browsers; if you just "trust that it works, without testing it" is better to use `useBuiltIns: "entry"`.

You can find more info at https://github.com/babel/babel-polyfills/#injection-methods (`useBuiltIns: "usage"` is an alias for `method: "usage-global"` of `babel-plugin-polyfill-corejs3`, while `useBuiltIns: "entry"` is an alias for `method: "entry-global"`).

**Reply by nyngwang** [reply]
@nicolo-ribaudo: Then why provide `usage` for us? It's misleading and error-prone! And even worse people have spent so much time on differentiating its usage with `entry`! By intuition `usage` sounds like a more wise choice(and thus people will prefer to choose it first). It can decrease the bundle size, which is the most important thing in the front-end. I don't believe that new-comers will all read the (verbose) comment I linked in my first comment.

I have to be the first one to tell you the truth about how we felt during the learning of Babel(, and usually Webpack altogether). Could you provide some "best practice" tutorials on the website so people will know "OK, those experts use this, I just first follow them then learn the details later".

**Reply by nicolo-ribaudo** [reply]
`usage` works in almost every case, I personally have never used `entry` :shrug:

I would gladly accept a PR to update the docs, mentioning that `useBuiltIns: usage` does not inject polyfills needed by your dependencies unless you are compiling them with Babel, and that https://github.com/babel/babel-polyfills/ might be a better option if your dependencies only need polyfills but no syntax transpilation because it lets you inject polyfills without transforming everything with `@babel/preset-env`.

**Reply by nyngwang** [reply]
> usage works in almost every case, I personally have never used entry 🤷
> [...]

@nicolo-ribaudo: (**Updated**) Great. This solves my problem :)

> I would gladly accept a PR to [...]

My pleasure. I will schedule a time for this.

> [...] and that https://github.com/babel/babel-polyfills/ might be a better option if [...]

If you're interested (probably not?) you will find that **I have starred and forked it hours ago** because I have planned to do exactly what you're saying.

And many thanks to you: Compared with those maintainers from Webpack, you're like 100x more responsive to answer people's (dumb?) questions with details.

**Reply by nicolo-ribaudo** [reply]
> My pleasure. I will schedule a time for this.

Thanks!

> And many thanks to you: Compared with those maintainers from Webpack, you're like 100x more responsive to answer people's (dumb?) questions with details.

Babel and Webpack are both big projects, and unfortunately it happens that in some periods either us or them don't have time to answer to everyone! What motivates me the most is that I'm not helping "just you", but potentially anyone who will find this public discussion in the future.

**Reply by icy0307** [reply]
Excuse me , I still get little bit confused about how library author should handle their polyfill and helper. Even with the new  [babel/babel-polyfills](https://github.com/babel/babel-polyfills) .
@nyngwang @nicolo-ribaudo 

## Question 1: What is the best practice for library maintainers to handle browser target?
An library author still not be able to know the browser target of its potential users. 
As you mentioned in [RFC: Rethink polyfilling story ](https://github.com/babel/babel/issues/10008#issue-446717469).
Its 2022 now, very few people need to load an Array.prototype.forEach polyfill. 
But as a library author, he or she has no choice.

## Question 2:  How should we polyfilling web API?
I know it is out of the scope of babel, But since core-js would not handle web API like `requestIdlecallback`.
What is the recommend way in the babel tool chains?

**Reply by nyngwang** [reply]
@icy0307 What confused you has nothing to do with how the two options were explained. Answering any of those will result in an off-topic thread. That is not appropriate.
